logicrunner/handle_executorresults.go

```diff
@@ -0,0 +1,201 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package logicrunner
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"go.opencensus.io/trace"
+
+	"github.com/insolar/insolar/insolar/flow"
+	"github.com/insolar/insolar/insolar/flow/bus"
+	"github.com/insolar/insolar/insolar/message"
+	"github.com/insolar/insolar/insolar/reply"
+	"github.com/insolar/insolar/instrumentation/inslogger"
+	"github.com/insolar/insolar/instrumentation/instracer"
+)
+
+// ---------------
+
+type initializeExecutionState struct {
+	LR  *LogicRunner
+	msg *message.ExecutorResults
+
+	Result struct {
+		es             *ExecutionState
+		clarifyPending bool
+	}
+}
+
+func (p *initializeExecutionState) Proceed(ctx context.Context) error {
+	logger := inslogger.FromContext(ctx)
+	ref := p.msg.GetReference()
+
+	state := p.LR.UpsertObjectState(ref)
+
+	state.Lock()
+	if state.ExecutionState == nil {
+		state.ExecutionState = &ExecutionState{
+			Ref:   ref,
+			Queue: make([]ExecutionQueueElement, 0),
+		}
+	}
+	es := state.ExecutionState
+	p.Result.es = es
+	state.Unlock()
+
+	p.Result.clarifyPending = false
+
+	es.Lock()
+
+	if es.pending == message.InPending {
+		if es.Current != nil {
+			logger.Debug("execution returned to node that is still executing pending")
+
+			es.pending = message.NotPending
+			es.PendingConfirmed = false
+		} else if p.msg.Pending == message.NotPending {
+			logger.Debug("executor we came to thinks that execution pending, but previous said to continue")
+
+			es.pending = message.NotPending
+		}
+	} else if es.pending == message.PendingUnknown {
+		es.pending = p.msg.Pending
+
+		if es.pending == message.PendingUnknown {
+			p.Result.clarifyPending = true
+		}
+	}
+
+	// set false to true is good, set true to false may be wrong, better make unnecessary call
+	if !es.LedgerHasMoreRequests && p.msg.LedgerHasMoreRequests {
+		es.LedgerHasMoreRequests = p.msg.LedgerHasMoreRequests
+	}
+
+	// prepare Queue
+	if p.msg.Queue != nil {
+		queueFromMessage := make([]ExecutionQueueElement, 0)
+		for _, qe := range p.msg.Queue {
+			queueFromMessage = append(
+				queueFromMessage,
+				ExecutionQueueElement{
+					ctx:     qe.Parcel.Context(context.Background()),
+					parcel:  qe.Parcel,
+					request: qe.Request,
+				})
+		}
+		es.Queue = append(queueFromMessage, es.Queue...)
+	}
+
+	es.Unlock()
+
+	return nil
+}
+
+// ---------------
+
+type HandleExecutorResults struct {
+	dep *Dependencies
+
+	Message bus.Message
+}
+
+func (h *HandleExecutorResults) realPresent(ctx context.Context, f flow.Flow) error {
+	parcel := h.Message.Parcel
+	logger := inslogger.FromContext(ctx)
+	msg := parcel.Message().(*message.ExecutorResults)
+
+	// now we have 2 different types of data in message.HandleExecutorResultsMessage
+	// one part of it is about consensus
+	// another one is about prepare state on new executor after pulse
+	// TODO make it in different goroutines
```

Comment:
Please don't left TODO/FIXME marks in the code. Create Jira issues if necessary.

Reply:
Copy past of previous code

Reply:
fixed